client/setec/store.go

```diff
@@ -349,22 +377,55 @@ func StaticFile(path string) (Secret, error) {
 	return func() []byte { return bs }, nil
 }
 
+// hasExpired reports whether cs is an undeclared secret whose last access time
+// was longer ago than the expiry window.
+func (s *Store) hasExpired(cs *cachedSecret) bool {
+	if cs.Declared {
+		return false // declared secrets do not expire
+	} else if s.expiryAge <= 0 {
+		return false // no expiry age is defined
+	}
+	age := time.Unix(s.timeNow(), 0).UTC().Sub(cs.lastAccessTime())
+	return age > s.expiryAge
+}
+
+// snapshotActive captures a point-in-time snapshot of the active names and
+// versions of all secrets known to the store. This permits an update poll to
+// do the time-consuming lookups outside the lock.
+func (s *Store) snapshotActive() map[string]api.SecretVersion {
+	s.active.RLock()
+	defer s.active.RUnlock()
+	m := make(map[string]api.SecretVersion)
+	for name, cs := range s.active.m {
+		if s.hasExpired(cs) {
+			m[name] = 0 // 0 means "expired"
+		} else {
+			m[name] = cs.Secret.Version
+		}
+	}
+	return m
+}
+
 // poll polls the service for the active version of each secret in s.active.m.
 // It populates updates with any secret values that have changed.
 func (s *Store) poll(ctx context.Context, updates map[string]*api.SecretValue) error {
-	s.active.RLock()
-	defer s.active.RUnlock()
 	var errs []error
-	for name, cv := range s.active.m {
-		sv, err := s.client.GetIfChanged(ctx, name, cv.Version)
+	for name, sv := range s.snapshotActive() {
+		// If the secret has expired, mark it for deletion.
+		if sv == 0 {
```

Comment:
Now here 0 turns into a nil, which feels quite weird. This feels like it wants to be doing a set intersection between updates and active, but the meaning of that is a little obscured.

Reply:
(addressed as noted above)